src/vm/dllimportcallback.cpp

```diff
@@ -33,6 +33,90 @@ struct UM2MThunk_Args
     int argLen;
 };
 
+class UMEntryThunkFreeList
+{
+public:
+    UMEntryThunkFreeList(size_t size) :
+        m_size(size),
+        m_count(0),
+        m_crst(CrstUMEntryThunkFreeList)
+    {
+        WRAPPER_NO_CONTRACT;
+    }
+
+    UMEntryThunk *GetUMEntryThunk()
+    {
+        WRAPPER_NO_CONTRACT;
+
+        if (m_count < m_size)
+            return NULL;
+
+        CrstHolder ch(&m_crst);
+
+        SListElem<UMEntryThunk*> *pElem = m_list.RemoveHead();
+
+        if (pElem == NULL)
+            return NULL;
+
+        --m_count;
+
+        return pElem->GetValue();
+    }
+
+    void AddToList(UMEntryThunk *pThunk)
+    {
+        WRAPPER_NO_CONTRACT;
+
+        CrstHolder ch(&m_crst);
+
+        if (m_count == m_size)
+        {
+            SListElem<UMEntryThunk*> *pElem = m_list.RemoveHead();
+
+            if (pElem != NULL)
+            {
+                UMEntryThunkFreeList::FreeThunk(pElem->GetValue());
+                delete pElem;
+            }
+        }
+        else
+        {
+            ++m_count;
+        }
+
+        m_list.InsertTail(new SListElem<UMEntryThunk*>(pThunk));
+    }
+
+    ~UMEntryThunkFreeList()
+    {
+        SListElem<UMEntryThunk*> *pElem = m_list.GetHead();
```

Comment:
This destructor won't be necessary once we stop allocating our own heap. (You do not need to worry about freeing memory allocated on LoaderHeap.)